File: server/src/main/java/ru/fullrest/mfr/server/configuration/GitConfiguration.java
package ru.fullrest.mfr.server.configuration;

import com.jcraft.jsch.JSch;
import com.jcraft.jsch.JSchException;
import com.jcraft.jsch.Session;
import org.eclipse.jgit.api.Git;
import org.eclipse.jgit.api.TransportConfigCallback;
import org.eclipse.jgit.lib.Repository;
import org.eclipse.jgit.storage.file.FileRepositoryBuilder;
import org.eclipse.jgit.transport.JschConfigSessionFactory;
import org.eclipse.jgit.transport.OpenSshConfig;
import org.eclipse.jgit.transport.SshSessionFactory;
import org.eclipse.jgit.transport.SshTransport;
import org.eclipse.jgit.util.FS;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.io.File;
import java.io.IOException;

@Configuration
public class GitConfiguration {
    private final static String GIT_FOLDER = ".git";

    @Value("${local.ssh-key}")
    private String sshKeyPath;

    @Value("${git.repository.local}")
    private String repositoryLink;

    @Bean
    public SshSessionFactory sshSessionFactory() {
        return new JschConfigSessionFactory() {
            @Override
            protected void configure(OpenSshConfig.Host hc, Session session) {
                session.setConfig("StrictHostKeyChecking", "no");
            }

            @Override
            protected JSch createDefaultJSch(FS fs) throws JSchException {
                JSch defaultJSch = super.createDefaultJSch(fs);
                defaultJSch.removeAllIdentity();
                defaultJSch.addIdentity(sshKeyPath);
                return defaultJSch;
            }
        };
    }

    @Bean
    public TransportConfigCallback transportConfigCallback(SshSessionFactory sshSessionFactory) {
        return transport -> {
            SshTransport sshTransport = (SshTransport) transport;
            sshTransport.setSshSessionFactory(sshSessionFactory);
        };
    }

    @Bean
    public Repository repository() throws IOException {
        return FileRepositoryBuilder.create(new File(repositoryLink + File.separator + GIT_FOLDER));
    }

    @Bean
    public Git git(Repository repository) {
        return new Git(repository);
    }
}

File: server/src/main/kotlin/com/lezenford/mfr/server/configuration/GitConfiguration.kt
package com.lezenford.mfr.server.configuration

import com.jcraft.jsch.JSch
import com.jcraft.jsch.Session
import com.lezenford.mfr.server.configuration.properties.ServerSettingProperties
import org.eclipse.jgit.api.TransportConfigCallback
import org.eclipse.jgit.transport.JschConfigSessionFactory
import org.eclipse.jgit.transport.OpenSshConfig
import org.eclipse.jgit.transport.SshSessionFactory
import org.eclipse.jgit.transport.SshTransport
import org.eclipse.jgit.util.FS
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class GitConfiguration(
    private val settingProperties: ServerSettingProperties
) {

    @Bean
    fun sshSessionFactory(): SshSessionFactory = object : JschConfigSessionFactory() {
        override fun configure(hc: OpenSshConfig.Host, session: Session) {
            session.setConfig("StrictHostKeyChecking", "no")
        }

        override fun createDefaultJSch(fs: FS): JSch =
            super.createDefaultJSch(fs).also {
                it.removeAllIdentity()
                it.addIdentity(settingProperties.build.key)
            }

    }

    @Bean
    fun transportConfigCallback(sshSessionFactory: SshSessionFactory): TransportConfigCallback =
        TransportConfigCallback { transport -> (transport as SshTransport).sshSessionFactory = sshSessionFactory }
}